test/e2e/node_reuse_test.go

```diff
@@ -0,0 +1,584 @@
+		}
+		deleting_count := 0
+		for _, machine := range machines.Items {
+			Expect(machine.Status.GetTypedPhase() == clusterv1.MachinePhaseProvisioning).To(BeFalse()) // Ensure no machine is provisioning
+			if machine.Status.GetTypedPhase() == clusterv1.MachinePhaseDeleting {
+				deleting_count++
+			}
+		}
+		if deleting_count != 1 {
+			return errors.New("Deleting machines count is not equal to 1")
+		}
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-machine-deleting")...).Should(Succeed())
+
+	Byf("Waiting until %d BMH(s) in deprovisioning state", NUMBER_OF_BMH_MD)
+	deprovisioning_bmh := []bmo.BareMetalHost{}
+	Eventually(
+		func() int {
+			bmhs := bmo.BareMetalHostList{}
+			Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+
+			for _, item := range bmhs.Items {
+				if item.Status.Provisioning.State == bmo.StateDeprovisioning {
+					deprovisioning_bmh = append(deprovisioning_bmh, item)
+				}
+			}
+			return len(deprovisioning_bmh)
+		},
+		e2eConfig.GetIntervals(specName, "wait-bmh-deprovisioning")...,
+	).Should(Equal(NUMBER_OF_BMH_MD), fmt.Sprintf("Deprovisioning bmhs are not equal to %d", NUMBER_OF_BMH_MD))
+
+	By("Waiting until above deprovisioning BMHs are in Ready state again")
+	Eventually(
+		func() error {
+			bmhs := bmo.BareMetalHostList{}
+			Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+			bmhConfMap := map[string]bmo.BareMetalHost{}
+			for _, v := range bmhs.Items {
+				bmhConfMap[v.Name] = v
+			}
+			for _, item := range deprovisioning_bmh {
+				if bmhConfMap[item.Name].Status.Provisioning.State != bmo.StateReady {
+					return errors.New("A bmh is not ready yet")
+				}
+			}
+			return nil
+		},
+		e2eConfig.GetIntervals(specName, "wait-bmh-deprovisioning-ready")...,
+	).Should(Succeed())
+
+	By("Checking if just deprovisioned became ready BMHs and re-used for next provisioning")
+	// TODO Probably we need to limit the number of provisioning to the previous save
+	Eventually(
+		func() error {
+			bmhs := bmo.BareMetalHostList{}
+			Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+			bmhConfMap := map[string]bmo.BareMetalHost{}
+			for _, v := range bmhs.Items {
+				bmhConfMap[v.Name] = v
+			}
+			for _, item := range deprovisioning_bmh {
+				if bmhConfMap[item.Name].Status.Provisioning.State != bmo.StateProvisioning {
+					return errors.New("A bmh is not provisioning yet")
+				}
+			}
+			return nil
+		},
+		e2eConfig.GetIntervals(specName, "wait-bmh-ready-provisioning")...,
+	).Should(Succeed())
+
+	Byf("Waiting until two machines become running and updated with the new %s k8s version", UPGRADED_K8S_VERSION)
+	Eventually(
+		func() error {
+			machines := &clusterv1.MachineList{}
+			err = targetClusterClient.List(ctx, machines, client.InNamespace(namespace))
+			fmt.Fprintf(GinkgoWriter, "Error:  %v\n", err)
```

Comment:
Use Logf()

Reply:
Should be use info logger / warning logger / error logger or just Logf() for everything ?

Reply:
Yeah, that is a good idea to use something else rather than Logf() to add categorize the messages. Also, please create a public function (like Logf) for error message only, so we have a common way of logging in e2e test framework.

Reply:
So far I have used only LogF I think having multi level logging needs touching files outside this context of this PR we can handle this later in separate one